nemo/collections/asr/parts/k2/utils.py

```diff
@@ -0,0 +1,205 @@
+    These supervisions are required for some k2 methods.
+    """
+    supervisions = torch.stack(
+        (torch.tensor(range(input_lengths.shape[0])), torch.zeros(input_lengths.shape[0]), input_lengths.cpu(),), 1,
+    ).to(torch.int32)
+    # the duration column has to be sorted in decreasing order
+    order = torch.argsort(supervisions[:, -1], descending=True)
+    return supervisions[order], order
+
+
+def invert_permutation(indices: torch.Tensor) -> torch.Tensor:
+    """Produces a tensor of reverse permutation for a given indices.
+    
+    Copied from https://github.com/k2-fsa/snowfall/blob/master/snowfall/common.py
+    """
+    ans = torch.zeros(indices.shape, device=indices.device, dtype=torch.long)
+    ans[indices] = torch.arange(0, indices.shape[0], device=indices.device)
+    return ans
+
+
+def make_blank_first(
+    blank_idx: int, log_probs: torch.Tensor, targets: torch.Tensor
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Puts blank logits at the first place in input log_probs tensor.
+    """
+    index = list(range(log_probs.shape[-1]))
+    del index[blank_idx]
+    index = torch.tensor([blank_idx] + index).to(log_probs.device)
+    new_log_probs = torch.index_select(log_probs, -1, index)
+    # TODO (alaptev): replace targets + 1 with torch.where to work for non-last blank_id
+    return new_log_probs, None if targets is None else targets + 1
+
+
+def load_graph(graph_path: str) -> k2.Fsa:
+    """Fsa graph loading helper function. Loads graphs stored in different formats.
+    """
+    if os.path.exists(graph_path):
+        errors = []
+        try:
+            graph_dict = torch.load(graph_path, map_location="cpu")
+            graph = k2.Fsa.from_dict(graph_dict)
+            return graph
+        except UnpicklingError as e:
+            errors.append(e)
+            with open(graph_path, "rt", encoding="utf-8") as f:
+                graph_txt = f.read()
+            for func, acceptor in itertools.product([k2.Fsa.from_str, k2.Fsa.from_openfst], [True, False]):
+                try:
+                    graph = func(graph_txt, acceptor=acceptor)
+                    return graph
+                except (TypeError, ValueError, RuntimeError) as e:
+                    errors.append(e)
+        raise Exception(errors)
+    else:
+        logging.warning(f"""No such file: '{graph_path}'""")
+        return None
+
+
+def intersect_with_self_loops(base_graph: k2.Fsa, aux_graph: k2.Fsa) -> k2.Fsa:
+    """Intersection helper function.
+    """
+    assert hasattr(base_graph, "aux_labels")
+    assert not hasattr(aux_graph, "aux_labels")
+    aux_graph_with_self_loops = k2.arc_sort(k2.add_epsilon_self_loops(aux_graph)).to(base_graph.device)
+    result = k2.intersect(k2.arc_sort(base_graph), aux_graph_with_self_loops, treat_epsilons_specially=False,)
+    setattr(result, "phones", result.labels)
+    return result
+
+
+def compose_with_self_loops(base_graph: k2.Fsa, aux_graph: k2.Fsa) -> k2.Fsa:
```

Comment:
Same as above

Reply:
Done in [7b2b4a4](https://github.com/NVIDIA/NeMo/pull/3351/commits/7b2b4a4ba4e34848d526d8c4ac53903bfbfe2fd2)